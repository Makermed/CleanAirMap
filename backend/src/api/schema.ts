import { makeSchema } from 'nexus'
import { join } from 'path'
import * as allTypes from './graphql'

export const schema = makeSchema({
  types: allTypes,
  outputs: {
    typegen: join(__dirname, '../..', 'nexus-typegen.ts'),
    schema: join(__dirname, '../..', 'schema.graphql'),
  },
  contextType: {
    module: join(__dirname, "./context.ts"),
    export: "Context",
  },
  sourceTypes: {
    headers: [ 'import { LocationModel } from "./src/dataAccess/dataTypes" '],
    modules: []
  }
})
