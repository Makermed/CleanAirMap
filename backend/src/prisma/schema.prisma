// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Location {
  locationId  Int      @id @default(autoincrement())
  name        String
  type        Int
  street      String
  locality    String?
  place       String
  district    String?
  region      String
  postcode    String?
  country     String
  latitude    Float
  longitude   Float
  description String?
  avgCo2      Float?
  rooms       Room[]
  created_id  Int
  created_at  DateTime @default(now())
}

model Room {
  roomId       Int       @id @default(autoincrement())
  locationId   Int
  location     Location  @relation(fields: [locationId], references: [locationId])
  name         String
  readings     Reading[]
  created_id   Int
  created_user User      @relation(fields: [created_id], references: [id])
  created_at   DateTime  @default(now())
}

model Reading {
  id           Int      @id @default(autoincrement())
  value        Decimal
  unit         String
  notes        String?
  locationId   Int
  readingRoom  Room?    @relation(fields: [roomId], references: [roomId])
  roomId       Int?
  created_id   Int
  created_user User     @relation(fields: [created_id], references: [id])
  created_at   DateTime @default(now())
}

model User {
  id       Int       @id @default(autoincrement())
  rooms    Room[]
  readings Reading[]
}
