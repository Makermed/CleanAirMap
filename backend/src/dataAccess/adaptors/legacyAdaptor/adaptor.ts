import { IDataAdaptor, ILocation, IReading, IRoom } from "../../interfaces";
import LocationDAO from "./LocationDAO";
import RoomDAO from "./RoomDAO";
import ReadingDAO from "./ReadingDAO";
import { HttpLink, ApolloClient, InMemoryCache, NormalizedCacheObject } from "@apollo/client/core";
import fetch from 'node-fetch'

/**
 * This adaptor queries from the legacy GraphQL endpoint and is just for
 * development/migration.
 */
export class LegacyAdaptor implements IDataAdaptor {
    client: ApolloClient<NormalizedCacheObject>;
    locationDAO: ILocation;
    roomDAO: IRoom;
    readingDAO: IReading;

    constructor(private endpoint: string) {
        this.client = new ApolloClient({
            link: new HttpLink({
                uri: endpoint,
                fetch: fetch
              }),
            cache: new InMemoryCache()
          });
      this.locationDAO = new LocationDAO(this.client);
      this.roomDAO = new RoomDAO(this.client);
      this.readingDAO = new ReadingDAO(this.client)
    }
  }
