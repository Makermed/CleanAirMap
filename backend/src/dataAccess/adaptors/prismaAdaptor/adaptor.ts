import { PrismaClient } from "@prisma/client";
import { IDataAdaptor, ILocation, IRoom } from "../../interfaces";
import { LocationDAO } from "./LocationDAO";
import { RoomDAO } from "./RoomDAO";

const prisma = new PrismaClient({
    log: ['query', 'info', 'warn', 'error'],
  })

export class PrismaDataAdaptor implements IDataAdaptor {
    client : PrismaClient
    locationDAO: ILocation
    roomDAO: IRoom

    constructor() {
        this.client = prisma
        this.locationDAO = new LocationDAO(this.client)
        this.roomDAO = new RoomDAO(this.client)
    }
}