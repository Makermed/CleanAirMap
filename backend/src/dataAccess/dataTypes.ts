export { User } from "@prisma/client"
export { Reading } from "@prisma/client"
export { Location } from "@prisma/client"
export { Room } from "@prisma/client"

// TODO: figure out if this can be referenced from Prisma types
export interface LocationModel {
    id: number
    name: String
    type: number
    street: String
    locality: String | null
    place: String
    district: String | null
    region: String
    postcode: String | null
    country: String
    latitude: number
    longitude: number
    description: String | null
    avgCo2: number | null
    created_id: number
    created_at: Date
}