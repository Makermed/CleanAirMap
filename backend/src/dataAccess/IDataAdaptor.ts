import { PrismaDataAdaptor } from "./adaptors/prismaAdaptor/adaptor";
import { ILocation, IReading, IRoom } from "./interfaces";

export interface IDataAdaptor {
    locationDAO : ILocation
    roomDAO : IRoom
    readingDAO : IReading
}