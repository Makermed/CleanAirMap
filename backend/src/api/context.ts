import { IDataAdaptor } from "../dataAccess/interfaces";
import { PrismaDataAdaptor } from "../dataAccess/adaptors/prismaAdaptor/adaptor";

export interface Context {
    db: IDataAdaptor
}

export const createContext =async () => ({
    db: new PrismaDataAdaptor()
})