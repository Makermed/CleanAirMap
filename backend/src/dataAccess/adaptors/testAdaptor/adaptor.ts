import { IDataAdaptor } from "../../IDataAdaptor";
import { ILocation } from "../../ILocation";
import { IRoom } from "../../IRoom";
import { LocationModel, RoomModel } from "../../dataTypes";

class TestLocationDAO implements ILocation {
    getById(id: number): Promise<LocationModel | null> {
        return Promise.reject(new Error("Not implemented"));
    }

    getMany(maxResults?: number | undefined): Promise<LocationModel[]> {
        return Promise.reject(new Error("Not implemented"));
    }

    getRooms(locationId: number): Promise<RoomModel[] | null> {
        return Promise.reject(new Error("Not implemented"));
    }

    create(reading: LocationModel): Promise<LocationModel | null> {
        return Promise.reject(new Error("Not implemented"));
    }
}

class TestRoomDAO implements IRoom {
    create(reading: RoomModel): Promise<RoomModel | null> {
        return Promise.reject(new Error("Not implemented"));
    }
    getById(id: number): Promise<RoomModel | null> {
        return Promise.reject(new Error("Not implemented"));
    }
    getByLocation(location: LocationModel): Promise<RoomModel[] | null> {
        return Promise.reject(new Error("Not implemented"));
    }
    getMany(maxResults?: number | undefined): Promise<RoomModel[]> {
        return Promise.reject(new Error("Not implemented"));
    }
}

export class TestAdaptor implements IDataAdaptor {
    locationDAO: ILocation;
    roomDAO: IRoom;

    constructor() {
        this.locationDAO = new TestLocationDAO();
        this.roomDAO = new TestRoomDAO();
    }
}