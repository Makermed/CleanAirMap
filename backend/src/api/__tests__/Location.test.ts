import { ApolloServer, GraphQLResponse } from "@apollo/server";
import { schema } from "../schema";
import { Context } from "../context"
import { TestAdaptor } from "../../dataAccess/adaptors/testAdaptor/adaptor";
import { LocationModel, RoomModel, LocationModelInput } from "../../dataAccess/dataTypes";
import { NexusGenInputs } from "../../../nexus-typegen";
import assert from 'assert';

const testServer = new ApolloServer<Context>({schema})

const testContext = {
    db: new TestAdaptor()
}

function mockSingleLocation() : Promise<LocationModel | null> {
    const location : LocationModel = {
        locationId: 1,
        name: 'Location 1',
        feature_type: 'poi',
        full_address: '123 Main St, AnyTown, StateProv, H0H0H0, Any Nation',
        geometry: { coordinates: [30,40.32908] },
        created_id: 1,
        created_at: new Date("2023-01-01")

    }
    return Promise.resolve(location)
}

function mockMultipleLocations() : Promise<LocationModel[] | null> {
  // generated by faker.js
  let locations: LocationModel[] = [{ locationId: 0, name: "Medhurst - Rau", feature_type: "poi", full_address: "685 Mueller Pines, South Eli, Saskatchewan, AX", geometry: { coordinates: [166.9186, -40.9219] }, created_id: 1, created_at: new Date("2022-10-07T21:53:43.633Z") }, { "locationId": 1, name: "Zieme - Reichel", feature_type: "poi", full_address: "113 Feil Cliffs, East Gilbert, Newfoundland and Labrador, NO", geometry: { coordinates: [-37.1637, 78.3971] }, created_id: 1, created_at: new Date("2023-01-29T08:04:08.914Z") }, { "locationId": 2, name: "Denesik - Balistreri", feature_type: "poi", full_address: "594 Helen Gateway, East   Connor, Prince Edward Island, MV", geometry: { coordinates: [-136.8178, -36.0622] }, created_id: 1, created_at: new Date("2022-12-21T04:16:55.300Z") }, { "locationId": 3, name: "Bernier - Quigley", feature_type: "poi", full_address: "315 Kertzmann Turnpike, East Alizeberg, New Brunswick, AU", geometry: { coordinates: [-146.3215, -10.6542] }, created_id: 1, created_at: new Date("2023-05-23T05:18:37.705Z") }, { "locationId": 4, name: "Tromp, Lockman and D'Amore", feature_type: "poi", full_address: "876 Emmerich Coves, Dorval, Alberta, PE", geometry: { coordinates: [-88.9264, 72.5722] }, created_id: 1, created_at: new Date("2022-11-14T09:32:51.352Z") }, { "locationId": 5, name: "Bednar Group", feature_type: "poi", full_address: "239 Borer Parkways, Port Fidel, Northwest Territories, KR", geometry: { coordinates: [-73.1518, 88.3478] }, created_id: 1, created_at: new Date("2022-05-18T00:52:33.174Z") }];

  return Promise.resolve(locations)
}

function mockSingleLocationRooms() : Promise<RoomModel[] | null> {
  const rooms : RoomModel[] = [{
    roomId : 1,
    locationId: 1,
    name: 'L1 Room 1',
    created_id: 1,
    created_at: new Date("2023-01-01")
  },{
    roomId : 2,
    locationId: 1,
    name: 'L1 Room 2',
    created_id: 1,
    created_at: new Date("2023-01-01")
  },{
    roomId : 3,
    locationId: 1,
    name: 'L1 Room 3',
    created_id: 1,
    created_at: new Date("2023-01-01")
  },]
  return Promise.resolve(rooms)
}

async function mockCreate(locIn: LocationModelInput) : Promise<LocationModel | null> {
  let out : LocationModel | null = await mockSingleLocation();
  if (out != null) {
    Object.assign(out, locIn)

    out.created_at = new Date('2023-01-05');
    out.created_id = 5;
  }

  return out;
}

describe("locations query field", () => {
  it('fetches a single location, no rooms', async() => {
    testContext.db.locationDAO.getById = jest.fn().mockReturnValueOnce(mockSingleLocation())

    const { body } : GraphQLResponse<Record<string,unknown>> = await testServer.executeOperation(
        {
            query: LOCATION_NO_ROOMS_QUERY,
            variables: { locationId: 1}
        },
        {
            contextValue: testContext
        }
    );

    assert(body.kind === 'single');
    expect(body.singleResult.errors).toBeUndefined();
    expect(body.singleResult.data).not.toBeNull();
    expect(body.singleResult.data?.locations).toMatchSnapshot();
  })

  it('fetches a single location\'s rooms', async() => {
    testContext.db.locationDAO.getById = jest.fn().mockReturnValueOnce(mockSingleLocation())
    testContext.db.locationDAO.getRooms = jest.fn().mockReturnValueOnce(mockSingleLocationRooms())

    const { body } = await testServer.executeOperation(
        {
            query: SINGLE_LOCATION_WITH_ROOMS_QUERY,
            variables: { locationId: 1}
        },
        {
            contextValue: testContext
        }
    )

    assert(body.kind === 'single');
    expect(body.singleResult.errors).toBeUndefined();
    expect(body.singleResult.data).not.toBeNull();
    expect(body.singleResult.data?.locations).toMatchSnapshot();
  })

  it('fetches multiple locations', async() => {
    testContext.db.locationDAO.getMany = jest.fn().mockReturnValueOnce(mockMultipleLocations())

    const { body } = await testServer.executeOperation(
        {
            query: LOCATION_NO_ROOMS_QUERY,
            variables: { locationId: null }
        },
        {
            contextValue: testContext
        }
    )

    assert(body.kind === 'single');
    expect(body.singleResult.errors).toBeUndefined();
    expect(body.singleResult.data).not.toBeNull();
    expect(body.singleResult.data?.locations).toMatchSnapshot();
  })

  it('stores a new location' , async () => {
    testContext.db.locationDAO.create = jest.fn().mockImplementationOnce(mockCreate);

    const locationIn : NexusGenInputs["LocationInputType"] = {
      type: 'Feature',
      geometry: { coordinates: [-86.9259644,43.4708553] },
      properties: {
        name: 'Create location test',
        feature_type: 'poi',
        full_address: '567 Kings Boul., SomePlace, Ontario, M4G 2E5, Canada'
      }
    };

    const { body } = await testServer.executeOperation(
      {
          query: CREATE_ONLY_LOCATION_QUERY,
          variables: { data: locationIn }
      },
      {
          contextValue: testContext
      }
    );

    assert(body.kind === 'single');
    expect(body.singleResult.errors).toBeUndefined();
    expect(body.singleResult.data).not.toBeNull();
    expect(body.singleResult.data?.createLocation).toMatchSnapshot();
  })
});

const LOCATION_NO_ROOMS_QUERY =
`query Locations($locationId: Int) {
  locations(locationId: $locationId) {
    type
    id
    geometry
    properties {
      name
      full_address
      feature_type
    }
  }
}`

const SINGLE_LOCATION_WITH_ROOMS_QUERY =
`query Locations($locationId: Int) {
  locations(locationId: $locationId) {
    id
    properties {
      name
    }
    rooms {
      roomId
      locationId
      name
      created_id
      created_at
    }
  }
}`

const CREATE_ONLY_LOCATION_QUERY =
`mutation CreateLocation($data: LocationInputType!) {
  createLocation(data: $data) {
    type
    id
    geometry
    properties {
      feature_type
      full_address
      name
    }
  }
}`