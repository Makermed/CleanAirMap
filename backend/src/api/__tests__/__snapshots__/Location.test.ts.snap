// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`locations query field fetches a single location's rooms 1`] = `
[
  {
    "id": 1,
    "properties": {
      "name": "Location 1",
    },
    "rooms": [
      {
        "created_at": 2023-01-01T00:00:00.000Z,
        "created_id": 1,
        "locationId": 1,
        "name": "L1 Room 1",
        "roomId": 1,
      },
      {
        "created_at": 2023-01-01T00:00:00.000Z,
        "created_id": 1,
        "locationId": 1,
        "name": "L1 Room 2",
        "roomId": 2,
      },
      {
        "created_at": 2023-01-01T00:00:00.000Z,
        "created_id": 1,
        "locationId": 1,
        "name": "L1 Room 3",
        "roomId": 3,
      },
    ],
  },
]
`;

exports[`locations query field fetches a single location, no rooms 1`] = `
[
  {
    "geometry": {
      "coordinates": [
        30,
        40.32908,
      ],
    },
    "id": 1,
    "properties": {
      "feature_type": "poi",
      "full_address": "123 Main St, AnyTown, StateProv, H0H0H0, Any Nation",
      "name": "Location 1",
    },
    "type": "Feature",
  },
]
`;

exports[`locations query field fetches multiple locations 1`] = `
[
  {
    "geometry": {
      "coordinates": [
        166.9186,
        -40.9219,
      ],
    },
    "id": 0,
    "properties": {
      "feature_type": "poi",
      "full_address": "685 Mueller Pines, South Eli, Saskatchewan, AX",
      "name": "Medhurst - Rau",
    },
    "type": "Feature",
  },
  {
    "geometry": {
      "coordinates": [
        -37.1637,
        78.3971,
      ],
    },
    "id": 1,
    "properties": {
      "feature_type": "poi",
      "full_address": "113 Feil Cliffs, East Gilbert, Newfoundland and Labrador, NO",
      "name": "Zieme - Reichel",
    },
    "type": "Feature",
  },
  {
    "geometry": {
      "coordinates": [
        -136.8178,
        -36.0622,
      ],
    },
    "id": 2,
    "properties": {
      "feature_type": "poi",
      "full_address": "594 Helen Gateway, East   Connor, Prince Edward Island, MV",
      "name": "Denesik - Balistreri",
    },
    "type": "Feature",
  },
  {
    "geometry": {
      "coordinates": [
        -146.3215,
        -10.6542,
      ],
    },
    "id": 3,
    "properties": {
      "feature_type": "poi",
      "full_address": "315 Kertzmann Turnpike, East Alizeberg, New Brunswick, AU",
      "name": "Bernier - Quigley",
    },
    "type": "Feature",
  },
  {
    "geometry": {
      "coordinates": [
        -88.9264,
        72.5722,
      ],
    },
    "id": 4,
    "properties": {
      "feature_type": "poi",
      "full_address": "876 Emmerich Coves, Dorval, Alberta, PE",
      "name": "Tromp, Lockman and D'Amore",
    },
    "type": "Feature",
  },
  {
    "geometry": {
      "coordinates": [
        -73.1518,
        88.3478,
      ],
    },
    "id": 5,
    "properties": {
      "feature_type": "poi",
      "full_address": "239 Borer Parkways, Port Fidel, Northwest Territories, KR",
      "name": "Bednar Group",
    },
    "type": "Feature",
  },
]
`;

exports[`locations query field stores a new location 1`] = `
{
  "geometry": {
    "coordinates": [
      -86.9259644,
      43.4708553,
    ],
  },
  "id": 1,
  "properties": {
    "feature_type": "poi",
    "full_address": "567 Kings Boul., SomePlace, Ontario, M4G 2E5, Canada",
    "name": "Create location test",
  },
  "type": "Feature",
}
`;
